modules/stochastic_tools/src/surrogates/GradientDescentTrainer.C

```diff
@@ -0,0 +1,146 @@
+#include "Sampler.h"
+#include "DenseMatrix.h"
+#include "MooseRandom.h"
+
+registerMooseObject("StochasticToolsApp", PolynomialLeastSquares);
+registerMooseObject("StochasticToolsApp", GradientDescentTrainer);
+
+InputParameters
+ObjectiveFunction::validParams()
+{
+  InputParameters params = GeneralUserObject::validParams();
+  return params;
+}
+
+ObjectiveFunction::ObjectiveFunction(const InputParameters & parameters)
+  : GeneralUserObject(parameters)
+{
+}
+
+InputParameters
+PolynomialLeastSquares::validParams()
+{
+  InputParameters params = ObjectiveFunction::validParams();
+  params.addParam<unsigned int>("order", 1, "The polynomial order to compute.");
+  params.addRequiredParam<VectorPostprocessorName>(
+      "vector_postprocessor",
+      "The name of the VectorPostprocessor object containing the training data.");
+  params.addRequiredParam<std::string>("x_vector",
+                                       "The name of the vector containing the x training data.");
+  params.addRequiredParam<std::string>("y_vector",
+                                       "The name of the vector containing the y training data.");
+  return params;
+}
+
+PolynomialLeastSquares::PolynomialLeastSquares(const InputParameters & parameters)
+  : ObjectiveFunction(parameters),
+    _order(getParam<unsigned int>("order")),
+    _x_training_data(getVectorPostprocessorValue(
+        "vector_postprocessor", getParam<std::string>("x_vector"), false)),
+    _y_training_data(getVectorPostprocessorValue(
+        "vector_postprocessor", getParam<std::string>("y_vector"), false))
+{
+}
+
+unsigned int
+PolynomialLeastSquares::size() const
+{
+  return _order + 1;
+}
+
+Real
+PolynomialLeastSquares::value(const DenseVector<Real> & /*x*/) const
+{
+  // TODO: Create a model to use this function and an AD version that computes the gradient
+  // automatically However, to use this object you don't need the training data. So, perhaps there
+  // just needs to be a PolynomialModel object that can accept the DenseVector values
+  return 0.0;
+}
+
+DenseVector<Real>
+PolynomialLeastSquares::gradient(const DenseVector<Real> & x) const
+{
+  // There is more Order of operation stuff between VPP and UO. In practice the ObjectFunctions
+  // can be created late via an action, for now hack it in
+  if (_b_vector.size() == 0)
+  {
+    _A_matrix.resize(_x_training_data.size(), _order + 1);
+    _b_vector = _y_training_data;
+    _Ax_minus_b.resize(_y_training_data.size());
+
+    // TODO: Loop over offsets and to a std::fill using _A.get_values() on the underlying
+    // std::vector
+    for (unsigned int col = 0; col < _order + 1; ++col)
+      for (unsigned int row = 0; row < _A_matrix.m(); ++row)
+        _A_matrix(row, col) = std::pow(_x_training_data[row], col);
+  }
+
+  // compute A^T(Ax-b)
+  _A_matrix.vector_mult(_Ax_minus_b, x);
+  _Ax_minus_b -= _b_vector;
+
+  DenseVector<Real> x_out;
+  _A_matrix.vector_mult_transpose(x_out, _Ax_minus_b);
+  return x_out;
```

Comment:
@fdkong Can PETSc do this sort of computation with a distributed dense matrix (https://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/Mat/MatCreateDense.html)?

Reply:
I will be surprised if it can not. But I did not try such things so far.

MatCreateDense

MatMultTranspose